File: python/hearttwin/schemas.py
"""Canonical Pydantic schemas for HeartTwin Lab.

All numeric values must carry source, unit, and confidence.
No value may be invented or hallucinated.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from uuid import uuid4

from pydantic import BaseModel, Field, model_validator

from python.hearttwin.safety import DISCLAIMER


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------


class OperatingMode(str, Enum):
    REST = "rest"
    MILD_ACTIVITY = "mild_activity"
    STRESS = "stress"
    RECOVERY = "recovery"


class RecoveryScenarioType(str, Enum):
    LOAD_REDUCTION = "load_reduction"
    OXYGEN_DELIVERY_IMPROVEMENT = "oxygen_delivery_improvement"
    CONTRACTILITY_SUPPORT = "contractility_support"
    CONDITIONING = "conditioning"
    STABILITY_MONITORING = "stability_monitoring"
    CUSTOM = "custom"


class TargetMetric(str, Enum):
    CARDIAC_OUTPUT = "cardiac_output"
    EF = "ef"
    PV_LOOP_EFFICIENCY = "pv_loop_efficiency"
    STABILITY = "stability"
    BALANCED = "balanced"


class DataUncertaintyPolicy(str, Enum):
    CONSERVATIVE = "conservative"
    MODERATE = "moderate"
    OPTIMISTIC = "optimistic"


class MissingValuePolicy(str, Enum):
    NULL = "null"
    PRIOR = "prior"
    REFUSE = "refuse"


class AgentStatus(str, Enum):
    SUCCESS = "success"
    WARNING = "warning"
    FAILED = "failed"


class SafetyLevel(str, Enum):
    CLEAR = "clear"
    CAUTION = "caution"
    BLOCKED = "blocked"


class ValueSource(str, Enum):
    FILE_EXTRACTION = "file_extraction"
    USER_INPUT = "user_input"
    DEFAULT_MODEL_PRIOR = "default_model_prior"
    DERIVED = "derived"


# ---------------------------------------------------------------------------
# Annotated measurement value with provenance
# ---------------------------------------------------------------------------


class MeasuredValue(BaseModel):
    value: float
    unit: str
    source: ValueSource
    confidence: float = Field(ge=0.0, le=1.0)
    source_file_id: Optional[str] = None
    method: Optional[str] = None
    evidence: Optional[str] = None


# ---------------------------------------------------------------------------
# Patient context (all optional — missing stays null)
# ---------------------------------------------------------------------------


class PatientContext(BaseModel):
    age_years: Optional[MeasuredValue] = None
    sex: Optional[str] = None
    height_cm: Optional[MeasuredValue] = None
    weight_kg: Optional[MeasuredValue] = None
    bsa_m2: Optional[MeasuredValue] = None
    notes: Optional[str] = None


# ---------------------------------------------------------------------------
# Clinical measurements
# ---------------------------------------------------------------------------


class Measurements(BaseModel):
    heart_rate_bpm: Optional[MeasuredValue] = None
    systolic_bp_mmhg: Optional[MeasuredValue] = None
    diastolic_bp_mmhg: Optional[MeasuredValue] = None
    edv_ml: Optional[MeasuredValue] = None
    esv_ml: Optional[MeasuredValue] = None
    ejection_fraction_pct: Optional[MeasuredValue] = None
    stroke_volume_ml: Optional[MeasuredValue] = None
    cardiac_output_l_min: Optional[MeasuredValue] = None
    troponin_ng_l: Optional[MeasuredValue] = None
    bnp_pg_ml: Optional[MeasuredValue] = None
    oxygen_saturation_pct: Optional[MeasuredValue] = None


# ---------------------------------------------------------------------------
# Electrophysiology
# ---------------------------------------------------------------------------


class Electrophysiology(BaseModel):
    rhythm_label: Optional[str] = None
    rr_interval_ms: Optional[MeasuredValue] = None
    qrs_duration_ms: Optional[MeasuredValue] = None
    qt_interval_ms: Optional[MeasuredValue] = None
    qtc_ms: Optional[MeasuredValue] = None
    r_peak_confidence: Optional[float] = Field(None, ge=0.0, le=1.0)
    conduction_delay_score: Optional[MeasuredValue] = None
    arrhythmia_instability_score: Optional[MeasuredValue] = None


# ---------------------------------------------------------------------------
# Hemodynamics indices
# ---------------------------------------------------------------------------


class Hemodynamics(BaseModel):
    preload_index: Optional[MeasuredValue] = None
    afterload_index: Optional[MeasuredValue] = None
    contractility_index: Optional[MeasuredValue] = None
    arterial_compliance_index: Optional[MeasuredValue] = None
    systemic_vascular_resistance_index: Optional[MeasuredValue] = None
    filling_pressure_index: Optional[MeasuredValue] = None
    pv_loop_area_index: Optional[MeasuredValue] = None


# ---------------------------------------------------------------------------
# Tissue state
# ---------------------------------------------------------------------------


class TissueState(BaseModel):
    scar_fraction: Optional[MeasuredValue] = None
    inflammation_index: Optional[MeasuredValue] = None
    oxygen_delivery_index: Optional[MeasuredValue] = None
    myocardial_oxygen_demand_index: Optional[MeasuredValue] = None
    stiffness_index: Optional[MeasuredValue] = None
    remodeling_index: Optional[MeasuredValue] = None
    damage_zone_location: Optional[str] = None


# ---------------------------------------------------------------------------
# Operating environment
# ---------------------------------------------------------------------------


class OperatingEnvironment(BaseModel):
    mode: OperatingMode = OperatingMode.REST
    simulation_duration_seconds: float = 60.0
    time_step_ms: float = 1.0
    activity_level_mets: float = 1.0
    hydration_index: float = 1.0
    sleep_recovery_index: float = 1.0
    stress_catecholamine_index: float = 1.0
    ambient_temperature_c: float = 22.0
    altitude_m: float = 0.0
    oxygen_fraction: float = 0.21
    medication_effect_profile: Optional[dict[str, float]] = None
    data_uncertainty_policy: DataUncertaintyPolicy = DataUncertaintyPolicy.MODERATE
    missing_value_policy: MissingValuePolicy = MissingValuePolicy.PRIOR


# ---------------------------------------------------------------------------
# Recovery / healing scenario config
# ---------------------------------------------------------------------------


class RecoveryConfig(BaseModel):
    recovery_horizon_days: int = Field(30, ge=1, le=365)
    scenario_type: RecoveryScenarioType = RecoveryScenarioType.LOAD_REDUCTION
    contractility_delta_per_day: float = 0.005
    afterload_delta_per_day: float = -0.005
    preload_delta_per_day: float = -0.003
    inflammation_decay_rate: float = 0.03
    oxygen_delivery_delta_per_day: float = 0.003
    stiffness_delta_per_day: float = -0.002
    scar_remodeling_rate: float = 0.001
    heart_rate_adaptation_rate: float = 0.002
    arrhythmia_stability_delta: float = 0.005
    max_safe_parameter_shift: float = 0.30
    uncertainty_penalty_weight: float = 0.2
    target_metric: TargetMetric = TargetMetric.BALANCED


# ---------------------------------------------------------------------------
# Simulation config (combines operating + recovery)
# ---------------------------------------------------------------------------


class SimulationConfig(BaseModel):
    operating: OperatingEnvironment = Field(default_factory=OperatingEnvironment)
    recovery: RecoveryConfig = Field(default_factory=RecoveryConfig)
    random_seed: int = 42


# ---------------------------------------------------------------------------
# Source map entry
# ---------------------------------------------------------------------------


class SourceMapEntry(BaseModel):
    field: str
    value: Optional[float] = None
    unit: str = ""
    source: ValueSource
    source_file_id: Optional[str] = None
    confidence: float
    method: Optional[str] = None
    evidence: Optional[str] = None


# ---------------------------------------------------------------------------
# Canonical cardiac twin state
# ---------------------------------------------------------------------------


class CardiacTwinState(BaseModel):
    case_id: str = Field(default_factory=lambda: str(uuid4()))
    created_at: datetime = Field(default_factory=datetime.utcnow)
    data_quality_score: float = Field(0.0, ge=0.0, le=1.0)
    safety_level: SafetyLevel = SafetyLevel.CLEAR
    patient_context: PatientContext = Field(default_factory=PatientContext)
    measurements: Measurements = Field(default_factory=Measurements)
    electrophysiology: Electrophysiology = Field(default_factory=Electrophysiology)
    hemodynamics: Hemodynamics = Field(default_factory=Hemodynamics)
    tissue_state: TissueState = Field(default_factory=TissueState)
    operating_environment: OperatingEnvironment = Field(default_factory=OperatingEnvironment)
    simulation_config: SimulationConfig = Field(default_factory=SimulationConfig)
    source_map: list[SourceMapEntry] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Agent response envelope
# ---------------------------------------------------------------------------


class AgentTraceStep(BaseModel):
    tool: str
    inputs: dict[str, Any]
    outputs: dict[str, Any]
    duration_ms: float


class AgentResponse(BaseModel):
    agent: str
    status: AgentStatus
    inputs_used: list[str] = Field(default_factory=list)
    outputs: dict[str, Any] = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    confidence: float = Field(0.0, ge=0.0, le=1.0)
    trace: list[AgentTraceStep] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# File metadata
# ---------------------------------------------------------------------------


class UploadedFile(BaseModel):
    file_id: str = Field(default_factory=lambda: str(uuid4()))
    filename: str
    content_type: str
    size_bytes: int
    storage_url: Optional[str] = None
    uploaded_at: datetime = Field(default_factory=datetime.utcnow)


# ---------------------------------------------------------------------------
# Case record
# ---------------------------------------------------------------------------


class CaseRecord(BaseModel):
    case_id: str = Field(default_factory=lambda: str(uuid4()))
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
    files: list[UploadedFile] = Field(default_factory=list)
    patient_notes: Optional[str] = None
    validated_fields: dict[str, Any] = Field(default_factory=dict)
    state: Optional[CardiacTwinState] = None
    stage_results: list[AgentResponse] = Field(default_factory=list)
    simulation_result: Optional[dict[str, Any]] = None
    recovery_scenarios: Optional[list[dict[str, Any]]] = None
    status: str = "created"
    safety_disclaimer: str = DISCLAIMER


# ---------------------------------------------------------------------------
# API request/response models
# ---------------------------------------------------------------------------


class CreateCaseRequest(BaseModel):
    patient_notes: Optional[str] = None
    simulation_config: Optional[SimulationConfig] = None


class ExtractRequest(BaseModel):
    file_ids: list[str]
    user_vitals: Optional[dict[str, Any]] = None


class OperateRequest(BaseModel):
    operating_environment: Optional[OperatingEnvironment] = None


class SimulateRecoveryRequest(BaseModel):
    recovery_config: Optional[RecoveryConfig] = None
    scenarios: Optional[list[RecoveryConfig]] = None


class HealthResponse(BaseModel):
    status: str = "ok"
    service: str = "hearttwin-api"
    disclaimer: str = DISCLAIMER
    environment: dict[str, Any] = Field(default_factory=dict)

File: web/lib/schemas.ts
import { z } from 'zod'

export const ValueSourceSchema = z.enum([
  'file_extraction',
  'user_input',
  'default_model_prior',
  'derived',
])

export const SafetyLevelSchema = z.enum(['clear', 'caution', 'blocked'])
export const AgentStatusSchema = z.enum(['success', 'warning', 'failed'])
export const OperatingModeSchema = z.enum(['rest', 'mild_activity', 'stress', 'recovery'])
export const DataUncertaintyPolicySchema = z.enum(['conservative', 'moderate', 'optimistic'])
export const MissingValuePolicySchema = z.enum(['null', 'prior', 'refuse'])

export const RecoveryScenarioTypeSchema = z.enum([
  'load_reduction',
  'oxygen_delivery_improvement',
  'contractility_support',
  'conditioning',
  'stability_monitoring',
  'custom',
])

export const TargetMetricSchema = z.enum([
  'cardiac_output',
  'ef',
  'pv_loop_efficiency',
  'stability',
  'balanced',
])

export const MeasuredValueSchema = z.object({
  value: z.number(),
  unit: z.string(),
  source: ValueSourceSchema,
  confidence: z.number().min(0).max(1),
  source_file_id: z.string().nullable().optional(),
  method: z.string().nullable().optional(),
  evidence: z.string().nullable().optional(),
})

export const PatientContextSchema = z.object({
  age_years: MeasuredValueSchema.nullable().optional(),
  sex: z.string().nullable().optional(),
  height_cm: MeasuredValueSchema.nullable().optional(),
  weight_kg: MeasuredValueSchema.nullable().optional(),
  bsa_m2: MeasuredValueSchema.nullable().optional(),
  notes: z.string().nullable().optional(),
})

export const MeasurementsSchema = z.object({
  heart_rate_bpm: MeasuredValueSchema.nullable().optional(),
  systolic_bp_mmhg: MeasuredValueSchema.nullable().optional(),
  diastolic_bp_mmhg: MeasuredValueSchema.nullable().optional(),
  edv_ml: MeasuredValueSchema.nullable().optional(),
  esv_ml: MeasuredValueSchema.nullable().optional(),
  ejection_fraction_pct: MeasuredValueSchema.nullable().optional(),
  stroke_volume_ml: MeasuredValueSchema.nullable().optional(),
  cardiac_output_l_min: MeasuredValueSchema.nullable().optional(),
  troponin_ng_l: MeasuredValueSchema.nullable().optional(),
  bnp_pg_ml: MeasuredValueSchema.nullable().optional(),
  oxygen_saturation_pct: MeasuredValueSchema.nullable().optional(),
})

export const ElectrophysiologySchema = z.object({
  rhythm_label: z.string().nullable().optional(),
  rr_interval_ms: MeasuredValueSchema.nullable().optional(),
  qrs_duration_ms: MeasuredValueSchema.nullable().optional(),
  qt_interval_ms: MeasuredValueSchema.nullable().optional(),
  qtc_ms: MeasuredValueSchema.nullable().optional(),
  r_peak_confidence: z.number().min(0).max(1).nullable().optional(),
  conduction_delay_score: MeasuredValueSchema.nullable().optional(),
  arrhythmia_instability_score: MeasuredValueSchema.nullable().optional(),
})

export const HemodynamicsSchema = z.object({
  preload_index: MeasuredValueSchema.nullable().optional(),
  afterload_index: MeasuredValueSchema.nullable().optional(),
  contractility_index: MeasuredValueSchema.nullable().optional(),
  arterial_compliance_index: MeasuredValueSchema.nullable().optional(),
  systemic_vascular_resistance_index: MeasuredValueSchema.nullable().optional(),
  filling_pressure_index: MeasuredValueSchema.nullable().optional(),
  pv_loop_area_index: MeasuredValueSchema.nullable().optional(),
})

export const TissueStateSchema = z.object({
  scar_fraction: MeasuredValueSchema.nullable().optional(),
  inflammation_index: MeasuredValueSchema.nullable().optional(),
  oxygen_delivery_index: MeasuredValueSchema.nullable().optional(),
  myocardial_oxygen_demand_index: MeasuredValueSchema.nullable().optional(),
  stiffness_index: MeasuredValueSchema.nullable().optional(),
  remodeling_index: MeasuredValueSchema.nullable().optional(),
  damage_zone_location: z.string().nullable().optional(),
})

export const OperatingEnvironmentSchema = z.object({
  mode: OperatingModeSchema,
  simulation_duration_seconds: z.number(),
  time_step_ms: z.number(),
  activity_level_mets: z.number(),
  hydration_index: z.number(),
  sleep_recovery_index: z.number(),
  stress_catecholamine_index: z.number(),
  ambient_temperature_c: z.number(),
  altitude_m: z.number(),
  oxygen_fraction: z.number(),
  medication_effect_profile: z.record(z.number()).nullable().optional(),
  data_uncertainty_policy: DataUncertaintyPolicySchema,
  missing_value_policy: MissingValuePolicySchema,
})

export const RecoveryConfigSchema = z.object({
  recovery_horizon_days: z.number().int().min(1).max(365),
  scenario_type: RecoveryScenarioTypeSchema,
  contractility_delta_per_day: z.number(),
  afterload_delta_per_day: z.number(),
  preload_delta_per_day: z.number(),
  inflammation_decay_rate: z.number(),
  oxygen_delivery_delta_per_day: z.number(),
  stiffness_delta_per_day: z.number(),
  scar_remodeling_rate: z.number(),
  heart_rate_adaptation_rate: z.number(),
  arrhythmia_stability_delta: z.number(),
  max_safe_parameter_shift: z.number(),
  uncertainty_penalty_weight: z.number(),
  target_metric: TargetMetricSchema,
})

export const SimulationConfigSchema = z.object({
  operating: OperatingEnvironmentSchema,
  recovery: RecoveryConfigSchema,
  random_seed: z.number().int(),
})

export const SourceMapEntrySchema = z.object({
  field: z.string(),
  value: z.number().nullable().optional(),
  unit: z.string(),
  source: ValueSourceSchema,
  source_file_id: z.string().nullable().optional(),
  confidence: z.number().min(0).max(1),
  method: z.string().nullable().optional(),
  evidence: z.string().nullable().optional(),
})

export const CardiacTwinStateSchema = z.object({
  case_id: z.string(),
  created_at: z.string(),
  data_quality_score: z.number().min(0).max(1),
  safety_level: SafetyLevelSchema,
  patient_context: PatientContextSchema,
  measurements: MeasurementsSchema,
  electrophysiology: ElectrophysiologySchema,
  hemodynamics: HemodynamicsSchema,
  tissue_state: TissueStateSchema,
  operating_environment: OperatingEnvironmentSchema,
  simulation_config: SimulationConfigSchema,
  source_map: z.array(SourceMapEntrySchema),
  warnings: z.array(z.string()),
})

export const AgentTraceStepSchema = z.object({
  tool: z.string(),
  inputs: z.record(z.unknown()),
  outputs: z.record(z.unknown()),
  duration_ms: z.number(),
})

export const AgentResponseSchema = z.object({
  agent: z.string(),
  status: AgentStatusSchema,
  inputs_used: z.array(z.string()),
  outputs: z.record(z.unknown()),
  warnings: z.array(z.string()),
  confidence: z.number().min(0).max(1),
  trace: z.array(AgentTraceStepSchema),
})

export const ManualVitalsSchema = z.object({
  heart_rate_bpm: z.number().min(20).max(280).optional(),
  systolic_bp_mmhg: z.number().min(50).max(260).optional(),
  diastolic_bp_mmhg: z.number().min(20).max(160).optional(),
  edv_ml: z.number().min(30).max(400).optional(),
  esv_ml: z.number().min(5).max(350).optional(),
  ejection_fraction_pct: z.number().min(5).max(90).optional(),
  troponin_ng_l: z.number().min(0).optional(),
  bnp_pg_ml: z.number().min(0).optional(),
  oxygen_saturation_pct: z.number().min(50).max(100).optional(),
  age_years: z.number().min(0).max(130).optional(),
  sex: z.enum(['male', 'female', 'other', '']).optional(),
  height_cm: z.number().min(50).max(250).optional(),
  weight_kg: z.number().min(1).max(500).optional(),
  notes: z.string().max(500).optional(),
}).refine(
  (data) => data.esv_ml == null || data.edv_ml == null || data.esv_ml < data.edv_ml,
  { message: 'ESV must be less than EDV', path: ['esv_ml'] },
).refine(
  (data) => (
    data.diastolic_bp_mmhg == null
    || data.systolic_bp_mmhg == null
    || data.diastolic_bp_mmhg < data.systolic_bp_mmhg
  ),
  { message: 'Diastolic BP must be less than systolic BP', path: ['diastolic_bp_mmhg'] },
)

export type ManualVitals = z.infer<typeof ManualVitalsSchema>

export function safeParse<T>(
  schema: z.ZodSchema<T>,
  data: unknown,
): { ok: true; data: T } | { ok: false; errors: z.ZodError } {
  const result = schema.safeParse(data)
  if (result.success) return { ok: true, data: result.data }
  return { ok: false, errors: result.error }
}
